### PR TITLE
DEVX-1979: Using callbacks in  producer reqeusts

### PR DESCRIPTION
**Pull request by bbejeck**
Covers https://github.com/confluentinc/kafka-tutorials/issues/406

### CODE REVIEW

File: _includes/tutorials/kafka-producer-application-callback/kafka/markup/dev/make-application.adoc [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+The `Callback` provides a way of handling any actions you want to take on request completion *_asynchronously_*.  Note that the `Callback` code executes on the producer's I/O thread so any actions here should execute quickly.  Any time consuming tasks could cause a delay in sending new records.
+
+The `KafkaProducer.send` method is asynchronous and returns as soon as the provided record is placed in the buffer of records to be sent to the broker. Once the broker acknowledges that the record has been appended to its log, the broker completes the produce request, which the application receives as `RecordMetadata`—information about the committed message.
+
+The code in the callback prints the `timestamp` and `offset` for each record sent using the `RecordMetadata` object.
+
+
+Now go ahead and create the following file at `src/main/java/io/confluent/developer/KafkaProducerCallbackApplication.java`.
```
Comment (bbejeck) [inline _includes/tutorials/kafka-producer-application-callback/kafka/markup/dev/make-application.adoc:80]: @ybyzek  this is only significant diff from PR #491 which you reviewed. Can you take a look at this?

File: _includes/tutorials/kafka-producer-application-callback/kafka/markup/dev/make-application.adoc [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+<5> Error handling portion-in this case printing the stacktrace to `System.err`
+
+The `KafkaProducerCallbackApplication.produce` method does some processing on a `String`, and then sends the https://kafka.apache.org/25/javadoc/org/apache/kafka/clients/producer/ProducerRecord.html[`ProducerRecord`].  While this code is a trivial example, it's enough to show the example of using a `KafkaProducer`.
+
+Notice that this overload of the https://javadoc.io/static/org.apache.kafka/kafka-clients/2.6.0/org/apache/kafka/clients/producer/KafkaProducer.html#send-org.apache.kafka.clients.producer.ProducerRecord-org.apache.kafka.clients.producer.Callback--[`KafkaProducer.send`] method accepts a second parameter, an instance of the https://kafka.apache.org/26/javadoc/org/apache/kafka/clients/producer/Callback.html[Callback] interface.
+
+The `Callback` provides a way of handling any actions you want to take on request completion *_asynchronously_*.  Note that the `Callback` code executes on the producer's I/O thread so any actions here should execute quickly.  Any time consuming tasks could cause a delay in sending new records.
```
Comment (ybyzek) [inline _includes/tutorials/kafka-producer-application-callback/kafka/markup/dev/make-application.adoc:73]:
```suggestion
The `Callback` provides a way of handling any actions you want to take on request completion *_asynchronously_*.  Note that the `Callback` code executes on the producer's I/O thread and any time consuming tasks could cause a delay in sending new records, so any code here should be designed to execute quickly.  
```

File: _includes/tutorials/kafka-producer-application-callback/kafka/markup/dev/make-application.adoc [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+The `KafkaProducer.send` method is asynchronous and returns as soon as the provided record is placed in the buffer of records to be sent to the broker. Once the broker acknowledges that the record has been appended to its log, the broker completes the produce request, which the application receives as `RecordMetadata`—information about the committed message.
+
+The code in the callback prints the `timestamp` and `offset` for each record sent using the `RecordMetadata` object.
```
Comment (ybyzek) [inline _includes/tutorials/kafka-producer-application-callback/kafka/markup/dev/make-application.adoc:77]:
```suggestion
In this example, the code in the callback just prints information from each record's `RecordMetadata` object, specifically the `timestamp` and `offset`.
```

File: _includes/tutorials/kafka-producer-application-callback/kafka/markup/dev/make-application.adoc [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+----
+
+<1> Process the String for sending message
+<2> Create the `ProducerRecord`
+<3> Send the record to the broker specifying a `Callback` instance as a lambda function
+<4> If there's no exceptions print the offset and timestamp of the acknowleged record
+<5> Error handling portion-in this case printing the stacktrace to `System.err`
+
+The `KafkaProducerCallbackApplication.produce` method does some processing on a `String`, and then sends the https://kafka.apache.org/25/javadoc/org/apache/kafka/clients/producer/ProducerRecord.html[`ProducerRecord`].  While this code is a trivial example, it's enough to show the example of using a `KafkaProducer`.
```
Comment (ybyzek) [inline _includes/tutorials/kafka-producer-application-callback/kafka/markup/dev/make-application.adoc:69]:
```suggestion
The `KafkaProducerCallbackApplication.produce` method does some processing on a `String`, and then sends it as a https://kafka.apache.org/25/javadoc/org/apache/kafka/clients/producer/ProducerRecord.html[`ProducerRecord`].  While this code is a trivial example, it's enough to show the example of using a `KafkaProducer`.
```

File: _includes/tutorials/kafka-producer-application-callback/kafka/markup/dev/make-application.adoc [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+Notice that this overload of the https://javadoc.io/static/org.apache.kafka/kafka-clients/2.6.0/org/apache/kafka/clients/producer/KafkaProducer.html#send-org.apache.kafka.clients.producer.ProducerRecord-org.apache.kafka.clients.producer.Callback--[`KafkaProducer.send`] method accepts a second parameter, an instance of the https://kafka.apache.org/26/javadoc/org/apache/kafka/clients/producer/Callback.html[Callback] interface.
+
+The `Callback` provides a way of handling any actions you want to take on request completion *_asynchronously_*.  Note that the `Callback` code executes on the producer's I/O thread so any actions here should execute quickly.  Any time consuming tasks could cause a delay in sending new records.
+
+The `KafkaProducer.send` method is asynchronous and returns as soon as the provided record is placed in the buffer of records to be sent to the broker. Once the broker acknowledges that the record has been appended to its log, the broker completes the produce request, which the application receives as `RecordMetadata`—information about the committed message.
```
Comment (ybyzek) [inline _includes/tutorials/kafka-producer-application-callback/kafka/markup/dev/make-application.adoc:75]: Is the send method itself technically `asynchronous`?  Or the callback is asynchronous?  Not sure if this is a distinction worth making, but FYC maybe tweak as:

```
The `KafkaProducer.send` method returns as soon as the provided record is placed in the buffer of records to be sent to the broker. Once the broker acknowledges that the record has been appended to its log, the broker completes the produce request by asynchronously sending back to the producer a `RecordMetadata` object—information about the committed message.
```

(or something like this)

Comment (bbejeck) [reply _includes/tutorials/kafka-producer-application-callback/kafka/markup/dev/make-application.adoc:75]: The send method itself is asynchronous. The callback is executed on the same thread that executes the request. What you have could work as well, but I think I'm going to leave this as is.

**Comment by bbejeck** [comment]
The relevant test for this PR passed, merging.

**Pull request merged**